### PR TITLE
TRD Fix mc 2 raw data generation

### PR DESCRIPTION
**Pull request by bazinski**
Parts of raw data generation are moved out of the trapsimulator dpl device.
The digits and tracklets files are now used to generate the raw data streams singular or per cru end point (72 files)
Raw data format now complies with what is coming out of CRU.
Digits are now included in addtion to the original of only tracklets.
Rejects any erroneous >3 tracklets.

This also includes an updated rawdata definitions for digits from the compressor branch.
Fix the optional groupings per file. Can now do per supermodule, per cru, per half cru, or all together.
Add instructions to README.md

### CODE REVIEW

**Comment by bazinski** [comment]
Dont bother reviewing yet, the halfchamber headers are wrong.

**Comment by martenole** [comment]
Hi Sean,
will you let us know when we should take a look?
Cheers,
Ole

**Comment by bazinski** [comment]
hi Ole
@martenole Just finished now moving the constants, so all done.
Cheers
Sean

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h
Comment (martenole) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: just for consistency: could we rename NUMBERXXX to NXXX, as we do it with the other constants for counts?

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h
Comment (martenole) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h]: very minor, but shouldn't it be std::ostream& operator<<(std::ostream& stream, const TrackletHCHeader*&* halfchamberheader);

File: Detectors/TRD/base/src/FeeParam.cxx
Comment (martenole) [inline Detectors/TRD/base/src/FeeParam.cxx]: Doesn't really matter, but you can also skip o2::trd::constants::

File: Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h
Comment (martenole) [inline Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h]: Could you add a comment what mFilePer is for instance?

File: Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h
Comment (martenole) [inline Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h:93]: Should gsl::span be used instead of std::vector for the input? Same for trigger records and tracklets.

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx]: rename to hcidOriginalDigits?

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx:188]: These loops are not needed if there is always only one entry in the tree, right?

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx:204]: The double loop I don't understand. Maybe digits trigger records should be the outer loop. And then for each time within the digit trigger records one can check if there are also tracklets connected to the same interaction (it can not be that there are tracklets but no digits). And in the case of tracklets then also convert the tracklet data. But it could in principle also be that there are digits but no tracklets, right? This case would now not be covered.

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx]: replace with o2::trd::constants::TIMEBINS?

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx]: As above?

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (martenole) [inline Detectors/TRD/simulation/src/Trap2CRU.cxx:469]: 21 -> o2::trd::constants::NADCMCM

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h
Comment (martenole) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h]: Not sure I understand it. The DigitHCHeader is not described in the tdp, right?

**Review by martenole — DISMISSED** [review]
That all looks good to me (as far as I can tell, but I did not go in very deep). Just a few minor comments in the code

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (bazinski) [reply Detectors/TRD/simulation/src/Trap2CRU.cxx:204]: there can be no digits if you choose not to use the digits in the conversion, as was the default before.
Even if the tracklets are empty the triggerrecord will be there but empty, it will still hit the digits. 
Yes I ummed and ahhhed which way to loop them.

For a system test, there is probably little reason to include the digits, they would get dumped before tracking in anycase (I think).

File: Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h:93]: probably

File: Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h]: oui

File: Detectors/TRD/base/src/FeeParam.cxx
Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx]: been burnt by namespace conflicts before hence the verbosity, it also makes it clearer where its coming from, can drop the o2::trd though.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h]: actually would not make a difference as TrackletHCHeader is a 32 bit integer, but will change.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: done

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h]: Its actually the only half chamber header defined in the TDP, and yip the citation is correct, although I see now the page number is actually 108, but that is likely to change, will take page number out.

File: Detectors/TRD/base/src/FeeParam.cxx
Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx]: done

File: Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/Trap2CRU.h]: done

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (bazinski) [reply Detectors/TRD/simulation/src/Trap2CRU.cxx]: done

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (bazinski) [reply Detectors/TRD/simulation/src/Trap2CRU.cxx]: done

**Comment by bazinski** [comment]
dont merge this until I sort out the bunchcrossing/orbit thing.

File: Detectors/TRD/simulation/src/Trap2CRU.cxx
Comment (bazinski) [reply Detectors/TRD/simulation/src/Trap2CRU.cxx:469]: done

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h
Comment (tdietel) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: I'm not sure if I see the latest version of the code here, but it seems the NUMBER... are still in. Also it looks like there is NHALFCRU and NUMBERHALFCRU.

Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: that is coming in a push at some point. I have a problem with 4 and 5 tracklets in an mcm ...
They are in the originating tracklets vector, and this messes up everything. in 200 events it happens about 5 or 6 times.
Still not sure why the same quadruplet of  det:283 Rob :4 mcm:9 row:10 appears twice with in 1 triggerrecord.
In this example, first it appears with 3 tracklets and then 80 tracklets later it appears with 1 tracklet. Tracklets 682/3/4 and 766 in https://murrays.web.cern.ch/trdtracklets.root or /eos/home/m/murrays/www/trdtracklets.root

Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]:
```
iTrap : 73 currdet : 425 Tracklets coming back from trapsimulator : 1  inserted at 766
det:283 Rob :4 mcm:9 row:10
tracklet : 766 hcid : col : row -- 566 : 1 : 10 pos 420
```
det is read from the mTrapsimulator[iTrap], currdet is the incoming function parameter for the detector to be processed.

just before and just after det was correctly 425. Only this one is being changed to 283 then the tracklets are sorted and this is "correctly" placed with the other 3 tracklets of 682 683 684 and resulting in a 4 tracklet mcm.

Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: The digits are definitely sorted correctly, there is no 283 in the middle of the 425s. 
In addition the trapsimulator is initialised correctly with 425 not 283 which is consistant with the digits being correct.
The complete sim/digitization/trapsim example in question  is https://murrays.web.cern.ch/problemrun.tar.gz

Comment (martenole) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: How did you generate that trdtracklets.root file? Because in there I also see that the tracklets are not correctly sorted by chamber ID. But I did a test simulation with the current dev branch for 200 events and in there the tracklets are sorted correctly

Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: std sim of 
o2-sim -n 200 -g pythia8 --skipModules ZDC
o2-sim-digitizer-workflow -b
o2-trd-trap-sim -b --trd-share-digits-manually

I put all the files up earlier this morning.
its rebased so the trapsim parts are the same as dev.

**Comment by bazinski** [comment]
Now rejects any instance where #tracklets>3.

**Comment by bazinski** [comment]
@martenole @tdietel @jolopezl you can go ahead and approve if you are happy, and assuming it gets through the CI

**Review by tdietel — DISMISSED** [review]
Looks ok, but we should understand where the extra tracklets come from. Do you have a todo list for those things?

**Comment by bazinski** [comment]
@tdietel yes the extra tracklets are understood and now fixed pending #5668 merging from @martenole today.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Constants.h]: just for completeness, this is fixed in #5668

**Review by jolopezl — DISMISSED** [review]
Looks good to me.

**Comment by bazinski** [comment]
This was approved previously, stuff was pushed to pass the ci. Can it be merged ? @shahor02 @sawenzel

**Pull request merged**
